fix(perf-test): stop after one nemesis cycle

soyacz opened the pull request:
During performance test with running nemesis, when nemesis ends
we want to prevent from starting another Nemesis cycle to not spoil
the results.
Also there's no need for running the load anymore.
This fix is about stopping nemesis in perf test after one cycle
and aborting load to finish the test quicker.

https://trello.com/c/SD5IN13F
## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

fruch commented on sdcm/cluster.py:
```diff
@@ -4667,8 +4667,9 @@ def kill_stress_thread(self):
         for loader in self.nodes:
             try:
-                loader.remoter.run(cmd='pgrep -f cassandra-stress | xargs -I{}  kill -TERM -{}',
+                loader.remoter.run(cmd='pgrep -f cassandra.stress | xargs -I{}  kill -TERM {}',
```
are you sure about those changes ? i.e. the process is called `cassandra-stress` and the minus there was for killing all the process tree. 

if it's not working we might want to use process group id, like mentioned here: 
https://stackoverflow.com/questions/392022/whats-the-best-way-to-send-a-signal-to-all-members-of-a-process-group

soyacz replied:
cassandra-stress is spawning another process, that has cassandra.stress. Somehow `pgrep -f` returns both processes (I tested it locally and it looks like `.` is a part of regex counted as any sign.). Minus was causing errors like  `kill: (-806457) - No such process`.
Also from recent build it looks that it works.

fruch commented on sdcm/cluster.py:
```diff
@@ -4677,6 +4678,7 @@ def kill_docker_loaders(self):
         for loader in self.nodes:
             try:
                 loader.remoter.run(cmd='docker ps -a -q | docker rm -f', verbose=False, ignore_status=True)
+                self.log.info("Killed docker loader on node: %s", loader.name)
```
please don't use info, we have enough print on this level, debug is good enough for this.

soyacz replied:
killing loaders is a very rare thing and important part of a test flow. I'd rather keep it at info level.

I agree, log info level is polluted - we should have some task to clean it up from unimportant things.

soyacz commented on sdcm/cluster.py:
```diff
@@ -4667,8 +4667,9 @@ def kill_stress_thread(self):
-                loader.remoter.run(cmd='pgrep -f cassandra-stress | xargs -I{}  kill -TERM -{}',
+                loader.remoter.run(cmd='pgrep -f cassandra.stress | xargs -I{}  kill -TERM {}',
                                    verbose=False, ignore_status=True)
+                self.log.info("Killed cassandra stress on node: %s", loader.name)
```
killing loaders is a very rare thing and important part of a test flow. I'd rather keep it at info level.

I agree, log info level is polluted - we should have some task to clean it up from unimportant things.

soyacz commented:
@fruch @fgelcer ping

fgelcer commented on performance_regression_test.py:
```diff
@@ -155,6 +158,14 @@ def _clean_email_data():
 
+    def _stop_load_after_one_nemesis_cycle(self):
+        time.sleep(300)  # wait 5 minutes to be sure nemesis has started
+        self.db_cluster.stop_nemesis(timeout=None)  # wait for Nemesis to end and don't start another cycle
```
does `timeout=None` mean that it will wait forever?

soyacz replied:
yes, if currently running nemesis does not end, this will be waiting forever.
But in that case `test_timeout` will break the process.

fgelcer replied:
yes, just wanted to be sure, that in this case, the timeout will be the test one, thanks

fgelcer reviewed (approved):
LGTM

fgelcer commented:
ping @aleksbykov @enaydanov

aleksbykov commented on test-cases/performance/perf-regression-latency-250gb-with-nemesis.yaml:
```diff
@@ -4,9 +4,9 @@ prepare_write_cmd: ["cassandra-stress write no-warmup cl=ALL n=62500000 -schema
-stress_cmd_w: "cassandra-stress write no-warmup cl=QUORUM duration=350m -schema 'replication(factor=3)' -port jmx=6868 -mode cql3 native -rate 'threads=50 throttle=5000/s' -col 'size=FIXED(128) n=FIXED(8)' -pop 'dist=gauss(1..250000000,125000000,12500000)' "
-stress_cmd_r: "cassandra-stress read no-warmup  cl=QUORUM duration=350m -schema 'replication(factor=3)' -port jmx=6868 -mode cql3 native -rate 'threads=50 throttle=4000/s' -col 'size=FIXED(128) n=FIXED(8)' -pop 'dist=gauss(1..250000000,125000000,12500000)' "
-stress_cmd_m: "cassandra-stress mixed no-warmup cl=QUORUM duration=350m -schema 'replication(factor=3)' -port jmx=6868 -mode cql3 native -rate 'threads=50 throttle=3500/s' -col 'size=FIXED(128) n=FIXED(8)' -pop 'dist=gauss(1..250000000,125000000,12500000)' "
+stress_cmd_w: "cassandra-stress write no-warmup cl=QUORUM duration=600m -schema 'replication(factor=3)' -port jmx=6868 -mode cql3 native -rate 'threads=50 throttle=5000/s' -col 'size=FIXED(128) n=FIXED(8)' -pop 'dist=gauss(1..250000000,125000000,12500000)' "
```
Why duration was increased?

soyacz replied:
we want stress thread to run till the end of nemesis. And because after nemesis ends we kill stress anyway, then we don't care about duration if is too long.

aleksbykov commented on performance_regression_test.py:
```diff
@@ -248,6 +259,7 @@ def run_workload(self, stress_cmd, nemesis=False):
             time.sleep(interval * 60)  # Sleeping one interval (in minutes) before starting the nemesis
             self.db_cluster.add_nemesis(nemesis=self.get_nemesis_class(), tester_obj=self)
             self.db_cluster.start_nemesis(interval=interval)
+            self._stop_load_after_one_nemesis_cycle()
         results = self.get_stress_results(queue=stress_queue)
         self.update_test_details(scrap_metrics_step=60)
```
Did you check, that c-s results properly parsed and collected?
i suspect, that when you kill the c-s commands, the c-s tool doesn't write summary to the log and in this case it will not properly parsed or don't parsed at all for current tests.

soyacz replied:
according to @fgelcer we don't need c-s tool summary in performance tests.

fgelcer replied:
@aleksbykov , IIRC in the latency with nemesis test, we don't do anything with this summary, do we?
maybe we just upload the results to ES, but it is not used to anything, as the latency values are calculated from Prometheus during each one of the operations..
maybe it can fail the final part, where we upload the data to ES, not sure about it...

aleksbykov reviewed (approved):
LGTM, 2 question only

soyacz commented:
@enaydanov @bentsi this is ready to be merged

soyacz commented:
ping @enaydanov @bentsi

fgelcer commented:
ping @enaydanov @fruch

roydahan commented:
@fgelcer & @soyacz this code doesn't cover the requirement of choosing the amount of cycles to run (its in the task description).

Moreover, it changes a test that is already running for long time and being tracked.
This should be changed back, the the amount of cycles should remain the same for the existing test and introduce a new test for one cycle (that may replace the existing test one day).

soyacz commented:
@roydahan as I understood, changed test was running only one nemesis cycle and load duration was tuned to match exactly one. This was causing issues as sometimes another cycle was starting and spoiling results. @fgelcer please correct me if I'm wrong.

As to setting multiple cycles, it needs to be fixed if this is required. We need to add which tests should have what settings for this. I've moved task back to queue.

fgelcer commented:
> @fgelcer & @soyacz this code doesn't cover the requirement of choosing the amount of cycles to run (its in the task description).
> 
> Moreover, it changes a test that is already running for long time and being tracked. This should be changed back, the the amount of cycles should remain the same for the existing test and introduce a new test for one cycle (that may replace the existing test one day).

sorry for the very late response.

but this is exactly what this test was meant to do, run 1 time the `disrupt_run_unique_sequence` nemesis, and stop...
we had yes few runs that started the 2nd cycle of `disrupt_run_unique_sequence`, but this was not the expected behavior, and they have misleading results.

AFAIU we have now the right behavior and the test is running correctly, and if we want to remove the "bad results", we should remove the ones that had the 2nd cycle started, and re-run with the same versions (or for now, just leave as is)

roydahan commented:
I think I confused it with something else.
When adding a node we are adding 3 nodes, and I confused it with cycles.
Please ignore my comment, although it would be nice to control the sequence with a parameter setting the number of sequences we want to repeat.

fgelcer commented:
> I think I confused it with something else. When adding a node we are adding 3 nodes, and I confused it with cycles. Please ignore my comment, although it would be nice to control the sequence with a parameter setting the number of sequences we want to repeat.

yes, @soyacz and I spoke about it, but at first, we just wanted it to do the right thing, and then we would enhance it with controllable amount of cycles... @soyacz , do we have a task for that enhancement? if not, would you open one please?

soyacz commented:
I added a task for it: https://trello.com/c/Xp2q7SfX
